content/en/docs/features/control-plane/entity-tag.md

```diff
@@ -0,0 +1,217 @@
+
+{{< note >}}
+
+**Collision example**: Consider a device that always sets the time to 1.1.1970 due to the lack of a real-time clock. This device has only one modifiable resource. Let's say the device is running with an updated resource that has an ETAG value of 1000, and the client also has the same ETAG value of 1000 cached. However, due to a power loss and subsequent restart (with no stored ETAGs in persistent storage), the random number generator returns 1, which is the same as the previous run. Meanwhile, another client updates the resource at the exact same time as the previous run but modifies a different property, resulting in an ETAG of 1000 for that resource.
+
+Now, the first client, unaware of the resource change by the other client, makes a request to the resource using the cached ETAG value of 1000 from the previous run. The device responds with a code of `VALID`, indicating that the ETAG is still valid. However, the first client is unaware that the resource has been changed by another client, and the device's resource actually has same the ETAG.
+
+{{< /note >}}
+
+```c
+static uint64_t g_etag;
+
+static uint64_t etag_random() {
+    return (oc_random_value()%1000)+1;
+}
+
+void oc_etag_init()
+{
+    g_etag = oc_clock_time() + etag_random();
+}
+
+// Get new ETAG for changed resource
+uint64_t oc_etag_get(void)
+{
+    uint64_t now = oc_clock_time();
+    if (now > g_etag)
+    {
+        g_etag = now;
+    }
+    g_etag = g_etag + etag_random();
+    return g_etag;
+}
+
+// Load ETAGs from storage and clear storage
+bool oc_etag_load_and_clear()
+{
+    uint64_t loaded_etags[OC_MAX_APP_RESOURCES];
+    bool ok = oc_etags_load(&loaded_etags);
+    oc_etag_clear_storage();
+    if (!ok)
+    {
+        return false;
+    }
+    uint64_t etag = oc_clock_time();
+    for (int i = 0; i < num_loaded_etag; ++i)
+    {
+        if (etag < loaded_etags[i])
+        {
+            etag = loaded_etags[i];
+        }
+    }
+    // Reinitialize global ETAG value if it can be less than loaded ETAGs
+    g_etag = etag + etag_random();
+
+    // Set ETAG for all resources
+    for (/*resources in device*/)
+    {
+        if (/*ETAG is in loaded_etags*/)
+        {
+            oc_resource_set_etag(resource, etag);
+        }
+        else
+        {
+            oc_resource_set_etag(resource, oc_etag_get());
+        }
+    }
+    return true;
+}
+```
+
+### Updating ETAG on the Device
+
+There are two methods for updating the ETAG when a resource changes:
+
+- `oc_notify_observers`: Notifies all observers that the resource has changed, what internally invokes `oc_resource_update_etag`.
+- `oc_resource_update_etag`: Updates the ETAG without notifying observers about the content change. This could be useful for non-observable resources.
```

Comment:
Can't there conceivably be a scenario in which you want to nofity observers but not update etag? I'm thinking whether we should tighly couple notifications of observers with etag, or keep them separate, or maybe creating something like `oc_resource_changed()` which would call everything that should be called on a resource change (notify observer and update etag for now, maybe more in the future)

Reply:
`oc_notify_observers`. Must be called only when resource has been changed. Ofcourse you can call it without any change but in this case it a bug :).

Reply:
_Should_ we put resource changed functionality into `oc_notify_observers` though? I would expect `oc_notify_observers` to notify observers ( 😃 ), but not to trigger everything that should happen on a resource change.

Reply:
fixed